The purpose of this file is to remove bounding boxes with area smaller than 40 pixel square 

In [ ]:
import json
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import cv2

# 1. Validation set

In [ ]:
df_validate = pd.read_csv('/content/drive/MyDrive/DL_Project/Data/validate.csv')
df_validate.head()

filename  width  height           class  xmin  ymin  xmax  ymax
0   21.png   1622     626          cam_re   822   314   844   336
1   21.png   1622     626  cam_dung_va_do   697   321   705   327
2   22.png   1622     626  cam_dung_va_do  1021   140  1161   270
3   47.png   1622     626     cam_con_lai   906   316   932   341
4   55.png   1622     626          cam_re  1370   225  1458   316

In [ ]:
df_validate['area'] = (df_validate['xmax'] - df_validate['xmin'])*(df_validate['ymax'] - df_validate['ymin'])

In [ ]:
df_validate.head()

filename  width  height           class  xmin  ymin  xmax  ymax   area
0   21.png   1622     626          cam_re   822   314   844   336    484
1   21.png   1622     626  cam_dung_va_do   697   321   705   327     48
2   22.png   1622     626  cam_dung_va_do  1021   140  1161   270  18200
3   47.png   1622     626     cam_con_lai   906   316   932   341    650
4   55.png   1622     626          cam_re  1370   225  1458   316   8008

In [ ]:
len(df_validate)

1287

In [ ]:
df_validate = df_validate[df_validate['area'] >= 40]
len(df_validate)

1208

In [ ]:
df_validate.drop(columns=['area'], inplace=True)
df_validate.head()

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:4174: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


filename  width  height           class  xmin  ymin  xmax  ymax
0   21.png   1622     626          cam_re   822   314   844   336
1   21.png   1622     626  cam_dung_va_do   697   321   705   327
2   22.png   1622     626  cam_dung_va_do  1021   140  1161   270
3   47.png   1622     626     cam_con_lai   906   316   932   341
4   55.png   1622     626          cam_re  1370   225  1458   316

In [ ]:
np.unique(df_validate['filename']).shape

(488,)

After removing 79 bboxes with area smaller than 40, there are 488 images corresponding to 1208 bboxes left in the validation set.


In [ ]:
df_validate.to_csv('/content/drive/MyDrive/DL_Project/Data/validate_40.csv', index=False)

# 2. Training set

In [ ]:
df_train = pd.read_csv('/content/drive/MyDrive/DL_Project/Data/train.csv')
df_train.head()

filename  width  height           class  xmin  ymin  xmax  ymax
0    3.png   1622     626  cam_dung_va_do   880   333   899   351
1    5.png   1622     626          cam_re  1069   355  1152   438
2    5.png   1622     626  cam_dung_va_do   768   480   777   487
3    6.png   1622     626       nguy_hiem   781   337   798   352
4   16.png   1622     626  cam_dung_va_do   733   352   740   360

In [ ]:
df_train['area'] = (df_train['xmax'] - df_train['xmin'])*(df_train['ymax'] - df_train['ymin'])

In [ ]:
len(df_train)

9713

In [ ]:
df_train.head()

filename  width  height           class  xmin  ymin  xmax  ymax  area
0    3.png   1622     626  cam_dung_va_do   880   333   899   351   342
1    5.png   1622     626          cam_re  1069   355  1152   438  6889
2    5.png   1622     626  cam_dung_va_do   768   480   777   487    63
3    6.png   1622     626       nguy_hiem   781   337   798   352   255
4   16.png   1622     626  cam_dung_va_do   733   352   740   360    56

In [ ]:
df_train = df_train[df_train['area'] >= 40]
len(df_train)

9023

In [ ]:
np.unique(df_train['filename']).shape

(3896,)

In [ ]:
df_train.to_csv('/content/drive/MyDrive/DL_Project/Data/train_40.csv', index=False)

# 3. Generate csv file for test set

We manually annotated the public set and converted the xml file to csv file


In [ ]:
pwd

'/content'

In [ ]:
!git clone https://github.com/tensorflow/models
!ls models

Cloning into 'models'...
remote: Enumerating objects: 45, done.
remote: Counting objects: 100% (45/45), done.
remote: Compressing objects: 100% (40/40), done.
remote: Total 50178 (delta 7), reused 21 (delta 1), pack-reused 50133
Receiving objects: 100% (50178/50178), 559.88 MiB | 31.89 MiB/s, done.
Resolving deltas: 100% (34127/34127), done.
AUTHORS     community	     ISSUES.md	official  README.md
CODEOWNERS  CONTRIBUTING.md  LICENSE	orbit	  research


In [ ]:
# Install protobuf
!wget -nc https://github.com/protocolbuffers/protobuf/releases/download/v3.14.0/protoc-3.14.0-linux-x86_64.zip
!unzip -n protoc-*.zip -d protoc

--2021-01-12 08:48:51--  https://github.com/protocolbuffers/protobuf/releases/download/v3.14.0/protoc-3.14.0-linux-x86_64.zip
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/23357588/40719200-25c2-11eb-8bf5-ac5ed2018051?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210112%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210112T084852Z&X-Amz-Expires=300&X-Amz-Signature=b5014b66a97a5f6d6aedef8529b787616be2237718f672f1fcfe8bb5e6ce5ae7&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=23357588&response-content-disposition=attachment%3B%20filename%3Dprotoc-3.14.0-linux-x86_64.zip&response-content-type=application%2Foctet-stream [following]
--2021-01-12 08:48:52--  https://github-production-release-asset-2e65be.s3.amazonaws.com/23357588/40719200-25c2-11eb-8bf5-ac5ed201805

In [ ]:
import os

# Add protoc binary to path (temporarily for this jupyter kernel only)
protoc_path = os.path.abspath('./protoc/bin')
if protoc_path not in os.environ['PATH']:
    os.environ['PATH'] += ':' + protoc_path

# verify
!protoc --version

libprotoc 3.0.0


In [ ]:
# :/
%cd models/research
!protoc object_detection/protos/*.proto --python_out=.
%cd ../..

/content/models/research
/content


In [ ]:
%cd models/research
!cp object_detection/packages/tf2/setup.py .
%pip install .
%cd ../..

  Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


/content


In [ ]:
# Test installation
# Warning: use !python3, do not use %run if you don't know exactly what it causes
!python3 models/research/object_detection/builders/model_builder_tf2_test.py

In [ ]:
!wget -nc https://tensorflow-object-detection-api-tutorial.readthedocs.io/en/latest/_downloads/da4babe668a8afb093cc7776d7e630f3/generate_tfrecord.py
%run generate_tfrecord.py --help

File ‘generate_tfrecord.py’ already there; not retrieving.

usage: generate_tfrecord.py [-h] [-x XML_DIR] [-l LABELS_PATH]
                            [-o OUTPUT_PATH] [-i IMAGE_DIR] [-c CSV_PATH]

Sample TensorFlow XML-to-TFRecord converter

optional arguments:
  -h, --help            show this help message and exit
  -x XML_DIR, --xml_dir XML_DIR
                        Path to the folder where the input .xml files are
                        stored.
  -l LABELS_PATH, --labels_path LABELS_PATH
                        Path to the labels (.pbtxt) file.
  -o OUTPUT_PATH, --output_path OUTPUT_PATH
                        Path of output TFRecord (.record) file.
  -i IMAGE_DIR, --image_dir IMAGE_DIR
                        Path to the folder where the input image files are
                        stored. Defaults to the same directory as XML_DIR.
  -c CSV_PATH, --csv_path CSV_PATH
                        Path of output .csv file. If none provided, then no
                        file will 

In [ ]:
xml_dir = '/content/drive/MyDrive/DL_Project/Data/Test_Set_Public'
csv_path = '/content/drive/MyDrive/DL_Project/Data/test.csv'
label_path = '/content/drive/MyDrive/DL_Project/Data/label_map.pbtxt'

In [ ]:
cd /content

/content


In [ ]:
!python3 generate_tfrecord.py --xml_dir /content/drive/MyDrive/DL_Project/Data/Test_Set_Public --labels_path /content/drive/MyDrive/DL_Project/Data/label_map.pbtxt  --output_path test.record --csv_path /content/drive/MyDrive/DL_Project/Data/test.csv

Successfully created the TFRecord file: test.record
Successfully created the CSV file: /content/drive/MyDrive/DL_Project/Data/test.csv


In [ ]:
import glob, os
files_xml = []
os.chdir("/content/drive/MyDrive/DL_Project/Data/Test_Set_Public")
for file in glob.glob("*.xml"):
    files_xml.append(file.split('.')[0])

In [ ]:
files_png = []
os.chdir("/content/drive/MyDrive/DL_Project/Data/Test_Set_Public")
for file in glob.glob("*.png"):
    files_png.append(file.split('.')[0])

In [ ]:
for i in range(len(files_png)):
  if files_png[i] not in files_xml:
    print(files_png[i])

12462
8124


In [ ]:
df_test = pd.read_csv('/content/drive/MyDrive/DL_Project/Data/test.csv')
df_test.head()

filename  width  height            class  xmin  ymin  xmax  ymax
0  12290.png   1622     626        hieu_lenh   610   362   614   366
1  12290.png   1622     626  cam_nguoc_chieu   610   358   615   362
2  12299.png   1622     626           cam_re   570   180   614   223
3  12299.png   1622     626        hieu_lenh   439   210   453   223
4  12297.png   1622     626        nguy_hiem   936   180   956   197

In [ ]:
df_test['area'] = (df_test['xmax'] - df_test['xmin'])*(df_test['ymax'] - df_test['ymin'])

In [ ]:
df_test.head()

filename  width  height            class  xmin  ymin  xmax  ymax  area
0  12290.png   1622     626        hieu_lenh   610   362   614   366    16
1  12290.png   1622     626  cam_nguoc_chieu   610   358   615   362    20
2  12299.png   1622     626           cam_re   570   180   614   223  1892
3  12299.png   1622     626        hieu_lenh   439   210   453   223   182
4  12297.png   1622     626        nguy_hiem   936   180   956   197   340

In [ ]:
len(df_test)

1191

In [ ]:
df_test = df_test[df_test['area'] >= 40]
len(df_test)

1159

In [ ]:
df_test.to_csv('/content/drive/MyDrive/DL_Project/Data/test_40.csv', index=False)